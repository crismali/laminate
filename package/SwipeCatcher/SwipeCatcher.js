import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { View, Dimensions, PanResponder, StyleSheet } from 'react-native'
import {
  LEFT_TO_RIGHT,
  RIGHT_TO_LEFT,
  TOP_TO_BOTTOM,
  BOTTOM_TO_TOP
} from './constants'

const styles = StyleSheet.create({
  container: {
    width: '100%',
    height: '100%',
    position: 'absolute'
  }
})

export class SwipeCatcher extends Component {
  static defaultProps = {
    direction: LEFT_TO_RIGHT
  }

  static propTypes = {
    children: PropTypes.any.isRequired,
    direction: PropTypes.oneOf([
      LEFT_TO_RIGHT,
      RIGHT_TO_LEFT,
      TOP_TO_BOTTOM,
      BOTTOM_TO_TOP
    ]),
    edgeThreshold: PropTypes.number,
    onGestureStart: PropTypes.func,
    onMove: PropTypes.func,
    onGestureRelease: PropTypes.func,
    onGestureTerminate: PropTypes.func,
    style: PropTypes.any
  }

  windowDimensions = () => Dimensions.get('window')

  invokeForCurrentDirection = (directions) => {
    const { direction } = this.props
    const result = directions[direction]
    if (result) {
      return result()
    } else {
      throw new Error(`Failed to handle direction '${direction}' in SwipeCatcher`)
    }
  }

  edgeThreshold = () => {
    const { edgeThreshold } = this.props
    const { height, width } = this.windowDimensions()

    if (edgeThreshold) {
      return edgeThreshold
    } else {
      return this.invokeForCurrentDirection({
        [LEFT_TO_RIGHT]: () => width,
        [RIGHT_TO_LEFT]: () => width,
        [TOP_TO_BOTTOM]: () => height,
        [BOTTOM_TO_TOP]: () => height
      })
    }
  }

  percentCompleteFromGestureState = ({ dx, dy }) => {
    const { height, width } = this.windowDimensions()

    const result = this.invokeForCurrentDirection({
      [LEFT_TO_RIGHT]: () => dx / width,
      [RIGHT_TO_LEFT]: () => -(dx / width),
      [TOP_TO_BOTTOM]: () => dy / height,
      [BOTTOM_TO_TOP]: () => -(dy / height)
    })

    return Math.max(result, 0)
  }

  velocityFromGestureState = ({ vx, vy }) => {
    return this.invokeForCurrentDirection({
      [LEFT_TO_RIGHT]: () => vx,
      [RIGHT_TO_LEFT]: () => vx,
      [TOP_TO_BOTTOM]: () => vy,
      [BOTTOM_TO_TOP]: () => vy
    })
  }

  dataFromGestureState = (gestureState) => {
    return {
      percentComplete: this.percentCompleteFromGestureState(gestureState),
      velocity: this.velocityFromGestureState(gestureState)
    }
  }

  panResponder = PanResponder.create({
    onMoveShouldSetPanResponderCapture: ({ nativeEvent }, gestureState) => {
      const { pageX, pageY } = nativeEvent
      const { dx, dy } = gestureState
      const edgeThreshold = this.edgeThreshold()

      return this.invokeForCurrentDirection({
        [LEFT_TO_RIGHT]: () => pageX <= edgeThreshold && dx > 0,
        [RIGHT_TO_LEFT]: () => pageX <= edgeThreshold && dx < 0,
        [TOP_TO_BOTTOM]: () => pageY <= edgeThreshold && dy > 0,
        [BOTTOM_TO_TOP]: () => pageY <= edgeThreshold && dy < 0
      })
    },

    onPanResponderGrant: () => {
      const { onGestureStart } = this.props
      onGestureStart && onGestureStart()
    },

    onPanResponderMove: (event, gestureState) => {
      const { onMove } = this.props
      if (onMove) {
        onMove(this.dataFromGestureState(gestureState))
      }
    },

    onPanResponderTerminate: (event, gestureState) => {
      const { onGestureTerminate } = this.props
      if (onGestureTerminate) {
        onGestureTerminate(this.dataFromGestureState(gestureState))
      }
    },

    onPanResponderRelease: (event, gestureState) => {
      const { onGestureRelease } = this.props
      if (onGestureRelease) {
        onGestureRelease(this.dataFromGestureState(gestureState))
      }
    }
  })

  render () {
    const { children, style } = this.props

    return <View style={[styles.container, style]} {...this.panResponder.panHandlers}>
      {children}
    </View>
  }
}
