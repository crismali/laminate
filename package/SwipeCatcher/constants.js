export const LEFT_TO_RIGHT = 'leftToRight'
export const RIGHT_TO_LEFT = 'rightToLeft'
export const TOP_TO_BOTTOM = 'topToBottom'
export const BOTTOM_TO_TOP = 'bottomToTop'
