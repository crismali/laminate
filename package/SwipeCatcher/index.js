import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { View, Dimensions, PanResponder, StyleSheet } from 'react-native'

const LEFT_TO_RIGHT = 'leftToRight'
const RIGHT_TO_LEFT = 'rightToLeft'
const TOP_TO_BOTTOM = 'topToBottom'
const BOTTOM_TO_TOP = 'bottomToTop'

const styles = StyleSheet.create({
  container: {
    width: '100%',
    height: '100%',
    position: 'absolute'
  }
})

export class SwipeCatcher extends Component {
  static defaultProps = {
    direction: LEFT_TO_RIGHT
  }

  static propTypes = {
    children: PropTypes.any.isRequired,
    direction: PropTypes.oneOf([
      LEFT_TO_RIGHT,
      RIGHT_TO_LEFT,
      TOP_TO_BOTTOM,
      BOTTOM_TO_TOP
    ]),
    edgeThreshold: PropTypes.number,
    onGestureStart: PropTypes.func,
    onMove: PropTypes.func,
    onGestureRelease: PropTypes.func,
    onGestureTerminate: PropTypes.func,
    style: PropTypes.any
  }

  windowDimensions = () => Dimensions.get('window')

  invokeForCurrentDirection = (directions) => {
    const { direction } = this.props
    const result = directions[direction]
    if (result) {
      return result()
    } else {
      throw new Error(`Failed to handle direction '${direction}' in SwipeCatcher`)
    }
  }

  edgeThreshold = () => {
    const { edgeThreshold } = this.props
    const { height, width } = this.windowDimensions()

    if (edgeThreshold) {
      return edgeThreshold
    } else {
      return this.invokeForCurrentDirection({
        leftToRight: () => width,
        rightToLeft: () => width,
        topToBottom: () => height,
        bottomToTop: () => height
      })
    }
  }

  percentCompleteFromGestureState = ({ dx, dy }) => {
    const { height, width } = this.windowDimensions()

    const result = this.invokeForCurrentDirection({
      leftToRight: () => dx / width,
      rightToLeft: () => 1 - dx / width,
      topToBottom: () => dy / height,
      bottomToTop: () => 1 - dy / height
    })

    return Math.max(result, 0)
  }

  velocityFromGestureState = ({ vx, vy }) => {
    return this.invokeForCurrentDirection({
      leftToRight: () => vx,
      rightToLeft: () => vx,
      topToBottom: () => vy,
      bottomToTop: () => vy
    })
  }

  dataFromGestureState = (gestureState) => {
    return {
      percentComplete: this.percentCompleteFromGestureState(gestureState),
      velocity: this.velocityFromGestureState(gestureState)
    }
  }

  panResponder = PanResponder.create({
    onMoveShouldSetPanResponderCapture: ({ nativeEvent }, gestureState) => {
      const { pageX, pageY } = nativeEvent
      const { dx, dy } = gestureState
      const edgeThreshold = this.edgeThreshold()

      return this.invokeForCurrentDirection({
        leftToRight: () => pageX <= edgeThreshold && dx > 0,
        rightToLeft: () => pageX <= edgeThreshold && dx < 0,
        topToBottom: () => pageY <= edgeThreshold && dy > 0,
        bottomToTop: () => pageY <= edgeThreshold && dy < 0
      })
    },

    onPanResponderGrant: () => {
      const { onGestureStart } = this.props
      onGestureStart && onGestureStart()
    },

    onPanResponderMove: (event, gestureState) => {
      const { onMove } = this.props
      if (onMove) {
        onMove(this.dataFromGestureState(gestureState))
      }
    },

    onPanResponderTerminate: (event, gestureState) => {
      const { onGestureTerminate } = this.props
      if (onGestureTerminate) {
        onGestureTerminate(this.dataFromGestureState(gestureState))
      }
    },

    onPanResponderRelease: (event, gestureState) => {
      const { onGestureRelease } = this.props
      if (onGestureRelease) {
        onGestureRelease(this.dataFromGestureState(gestureState))
      }
    }
  })

  render () {
    const { children, style } = this.props

    return <View style={[styles.container, style]} {...this.panResponder.panHandlers}>
      {children}
    </View>
  }
}
