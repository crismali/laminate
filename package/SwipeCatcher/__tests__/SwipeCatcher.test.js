/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import { SwipeCatcher } from '../SwipeCatcher'
import {
  LEFT_TO_RIGHT,
  RIGHT_TO_LEFT,
  TOP_TO_BOTTOM,
  BOTTOM_TO_TOP
} from '../constants'

const Child = () => 'a child'

describe('#edgeThreshold', () => {
  describe('when given as a prop', () => {
    it('is the prop', () => {
      const subject = shallow(<SwipeCatcher edgeThreshold={54}><Child/></SwipeCatcher>)
      expect(subject.instance().edgeThreshold()).toEqual(54)
    })
  })

  describe('when given a horizontal direction', () => {
    it('is the window width', () => {
      const subject = shallow(<SwipeCatcher direction={LEFT_TO_RIGHT}><Child/></SwipeCatcher>)
      expect(subject.instance().edgeThreshold()).toEqual(375)
      subject.setProps({ direction: RIGHT_TO_LEFT })
      expect(subject.instance().edgeThreshold()).toEqual(375)
    })
  })

  describe('when given a vertical direction', () => {
    it('is the window height', () => {
      const subject = shallow(<SwipeCatcher direction={TOP_TO_BOTTOM}><Child/></SwipeCatcher>)
      expect(subject.instance().edgeThreshold()).toEqual(667)
      subject.setProps({ direction: BOTTOM_TO_TOP })
      expect(subject.instance().edgeThreshold()).toEqual(667)
    })
  })
})

describe('#percentCompleteFromGestureState', () => {
  it('is calculated properly when left to right', () => {
    const subject = shallow(<SwipeCatcher direction={LEFT_TO_RIGHT}><Child/></SwipeCatcher>)
    const result = subject.instance().percentCompleteFromGestureState({ dx: 75, dy: 88 })
    expect(result).toEqual(0.2)
  })

  it('is calculated properly when right to left', () => {
    const subject = shallow(<SwipeCatcher direction={RIGHT_TO_LEFT}><Child/></SwipeCatcher>)
    const result = subject.instance().percentCompleteFromGestureState({ dx: -75, dy: 88 })
    expect(result).toEqual(0.2)
  })

  it('is calculated properly when top to bottom', () => {
    const subject = shallow(<SwipeCatcher direction={TOP_TO_BOTTOM}><Child/></SwipeCatcher>)
    const result = subject.instance().percentCompleteFromGestureState({ dx: 88, dy: 133.4 })
    expect(result).toEqual(0.2)
  })

  it('is calculated properly when bottom to top', () => {
    const subject = shallow(<SwipeCatcher direction={BOTTOM_TO_TOP}><Child/></SwipeCatcher>)
    const result = subject.instance().percentCompleteFromGestureState({ dx: 88, dy: -133.4 })
    expect(result).toEqual(0.2)
  })
})

describe('#velocityFromGestureState', () => {
  describe('when given a horizontal direction', () => {
    it('is the vx', () => {
      const subject = shallow(<SwipeCatcher direction={LEFT_TO_RIGHT}><Child/></SwipeCatcher>)
      expect(subject.instance().velocityFromGestureState({ vx: 12, vy: 13 })).toEqual(12)
      subject.setProps({ direction: RIGHT_TO_LEFT })
      expect(subject.instance().velocityFromGestureState({ vx: 12, vy: 13 })).toEqual(12)
    })
  })

  describe('when given a vertical direction', () => {
    it('is the vy', () => {
      const subject = shallow(<SwipeCatcher direction={TOP_TO_BOTTOM}><Child/></SwipeCatcher>)
      expect(subject.instance().velocityFromGestureState({ vx: 12, vy: 13 })).toEqual(13)
      subject.setProps({ direction: BOTTOM_TO_TOP })
      expect(subject.instance().velocityFromGestureState({ vx: 12, vy: 13 })).toEqual(13)
    })
  })
})

describe('#dataFromGestureState', () => {
  it('is the percentComplete and velocity from the gesture', () => {
    const subject = shallow(<SwipeCatcher direction={LEFT_TO_RIGHT}><Child/></SwipeCatcher>)
    const result = subject.instance().dataFromGestureState({ dx: 75, vx: 13 })
    expect(result).toEqual({ percentComplete: 0.2, velocity: 13 })
  })
})

describe('#onMoveShouldSetPanResponderCapture', () => {
  describe('when left to right', () => {
    it('is true when the pageX is less than the edge threshold and the dx is greater than 0', () => {
      const subject = shallow(<SwipeCatcher direction={LEFT_TO_RIGHT} edgeThreshold={50}><Child/></SwipeCatcher>)
      const { onMoveShouldSetPanResponderCapture } = subject.instance().panResponder.panHandlers
      expect(onMoveShouldSetPanResponderCapture({ nativeEvent: { pageX: 51 } }, { dx: 0 })).toEqual(false)
      expect(onMoveShouldSetPanResponderCapture({ nativeEvent: { pageX: 49 } }, { dx: 0 })).toEqual(false)
      expect(onMoveShouldSetPanResponderCapture({ nativeEvent: { pageX: 51 } }, { dx: 3 })).toEqual(false)
      expect(onMoveShouldSetPanResponderCapture({ nativeEvent: { pageX: 49 } }, { dx: 3 })).toEqual(true)
    })
  })

  describe('when right to left', () => {
    it('is true when the pageX is less than the edge threshold and the dx is less than 0', () => {
      const subject = shallow(<SwipeCatcher direction={RIGHT_TO_LEFT} edgeThreshold={50}><Child/></SwipeCatcher>)
      const { onMoveShouldSetPanResponderCapture } = subject.instance().panResponder.panHandlers
      expect(onMoveShouldSetPanResponderCapture({ nativeEvent: { pageX: 51 } }, { dx: 0 })).toEqual(false)
      expect(onMoveShouldSetPanResponderCapture({ nativeEvent: { pageX: 49 } }, { dx: 0 })).toEqual(false)
      expect(onMoveShouldSetPanResponderCapture({ nativeEvent: { pageX: 51 } }, { dx: -3 })).toEqual(false)
      expect(onMoveShouldSetPanResponderCapture({ nativeEvent: { pageX: 49 } }, { dx: -3 })).toEqual(true)
    })
  })

  describe('when top to bottom', () => {
    it('is true when the pageY is less than the edge threshold and the dy is greater than 0', () => {
      const subject = shallow(<SwipeCatcher direction={TOP_TO_BOTTOM} edgeThreshold={50}><Child/></SwipeCatcher>)
      const { onMoveShouldSetPanResponderCapture } = subject.instance().panResponder.panHandlers
      expect(onMoveShouldSetPanResponderCapture({ nativeEvent: { pageY: 51 } }, { dy: 0 })).toEqual(false)
      expect(onMoveShouldSetPanResponderCapture({ nativeEvent: { pageY: 49 } }, { dy: 0 })).toEqual(false)
      expect(onMoveShouldSetPanResponderCapture({ nativeEvent: { pageY: 51 } }, { dy: 3 })).toEqual(false)
      expect(onMoveShouldSetPanResponderCapture({ nativeEvent: { pageY: 49 } }, { dy: 3 })).toEqual(true)
    })
  })

  describe('when bottom to top', () => {
    it('is true when the pageY is less than the edge threshold and the dy is less than 0', () => {
      const subject = shallow(<SwipeCatcher direction={BOTTOM_TO_TOP} edgeThreshold={50}><Child/></SwipeCatcher>)
      const { onMoveShouldSetPanResponderCapture } = subject.instance().panResponder.panHandlers
      expect(onMoveShouldSetPanResponderCapture({ nativeEvent: { pageY: 51 } }, { dy: 0 })).toEqual(false)
      expect(onMoveShouldSetPanResponderCapture({ nativeEvent: { pageY: 49 } }, { dy: 0 })).toEqual(false)
      expect(onMoveShouldSetPanResponderCapture({ nativeEvent: { pageY: 51 } }, { dy: -3 })).toEqual(false)
      expect(onMoveShouldSetPanResponderCapture({ nativeEvent: { pageY: 49 } }, { dy: -3 })).toEqual(true)
    })
  })
})

describe('#onPanResponderGrant', () => {
  it('calls the onGestureStart callback if given', () => {
    const subject = shallow(<SwipeCatcher><Child/></SwipeCatcher>)
    const { onPanResponderGrant } = subject.instance().panResponder.panHandlers
    expect(onPanResponderGrant).not.toThrow()
    const onGestureStart = jest.fn()
    subject.setProps({ onGestureStart })
    onPanResponderGrant()
    expect(onGestureStart).toHaveBeenCalled()
  })
})

describe('#onPanResponderMove', () => {
  it('calls onMove when given', () => {
    const subject = shallow(<SwipeCatcher><Child/></SwipeCatcher>)
    const { onPanResponderMove } = subject.instance().panResponder.panHandlers
    const gestureState = { vx: 5, dx: 75 }
    expect(() => onPanResponderMove({}, gestureState)).not.toThrow()
    const onMove = jest.fn()
    subject.setProps({ onMove })
    onPanResponderMove({}, gestureState)
    expect(onMove).toHaveBeenCalledWith(subject.instance().dataFromGestureState(gestureState))
  })
})

describe('#onPanResponderTerminate', () => {
  it('calls onGestureTerminate when given', () => {
    const subject = shallow(<SwipeCatcher><Child/></SwipeCatcher>)
    const { onPanResponderTerminate } = subject.instance().panResponder.panHandlers
    const gestureState = { vx: 5, dx: 75 }
    expect(() => onPanResponderTerminate({}, gestureState)).not.toThrow()
    const onGestureTerminate = jest.fn()
    subject.setProps({ onGestureTerminate })
    onPanResponderTerminate({}, gestureState)
    expect(onGestureTerminate).toHaveBeenCalledWith(subject.instance().dataFromGestureState(gestureState))
  })
})

describe('#onPanResponderRelease', () => {
  it('calls onPanResponderRelease when given', () => {
    const subject = shallow(<SwipeCatcher><Child/></SwipeCatcher>)
    const { onPanResponderRelease } = subject.instance().panResponder.panHandlers
    const gestureState = { vx: 5, dx: 75 }
    expect(() => onPanResponderRelease({}, gestureState)).not.toThrow()
    const onGestureRelease = jest.fn()
    subject.setProps({ onGestureRelease })
    onPanResponderRelease({}, gestureState)
    expect(onGestureRelease).toHaveBeenCalledWith(subject.instance().dataFromGestureState(gestureState))
  })
})

describe('#render', () => {
  it('renders properly', () => {
    const subject = shallow(<SwipeCatcher style={{ inline: 'style' }}><Child/></SwipeCatcher>)
    expect(subject).toMatchSnapshot()
  })
})
