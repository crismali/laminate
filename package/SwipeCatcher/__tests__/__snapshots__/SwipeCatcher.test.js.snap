// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render renders properly 1`] = `
<View
  onMoveShouldSetPanResponderCapture={[Function]}
  onPanResponderGrant={[Function]}
  onPanResponderMove={[Function]}
  onPanResponderRelease={[Function]}
  onPanResponderTerminate={[Function]}
  style={
    Array [
      Object {
        "height": "100%",
        "position": "absolute",
        "width": "100%",
      },
      Object {
        "inline": "style",
      },
    ]
  }
>
  <Child />
</View>
`;
